Sources/SourceKitLSP/DocumentManager.swift

```diff
@@ -24,6 +25,13 @@ public struct DocumentSnapshot {
   /// the tokens are updated independently and only used internally.
   public var tokens: DocumentTokens
 
+  /// This information is used to determine whether a syntax node can be re-used in incremental parsing.
+  ///
+  /// The property is not nil only after the document is parsed.
+  public var lookaheadRanges: LookaheadRanges? {
+    document.latestLookaheadRanges
+  }
+
```

Comment:
We also need to store the `LookaheadRanges` in the `DocumentSnapshot`. At the moment, you could get into the following race condition: 

- We get a document snapshot and which stores its syntax tree (in `DocumentSnapshot.tokens` in `SwiftLanguageServer.changeDocument`)
- Now an edit is processed, which modifies the `Document` that’s underlying the `DocumentSnapshot`
- If we access `snapshot.lookaheadRanges`, we now get the `LookaheadRanges` from the updated document, which is out-of-sync with the syntax tree stored in the `DocumentSnapshot` and the text stored in the `DocumentSnapshot`

I would just store the `LookaheadRanges` in `DocumentTokens`. That way it’s stored right next to the syntax tree they belong to and you also don’t need the `updateLookaheadRanges` method to update them independently of the syntax tree.

---

Edit: On second thought, the race condition won’t actually happen because `updateSyntaxTree` is always executed on `SwiftLangaugeServer.queue` but still, we shouldn’t be relying on that invariant.

Reply:
I think storing `lookaheadRanges` in `DocumentTokens` is a good idea. 👍